source/module.c

```diff
@@ -931,6 +939,40 @@ static void s_handle_failed_callback(napi_env env, napi_value function, napi_sta
     }
 }
 
+napi_status aws_napi_create_external_arraybuffer(
+    napi_env env,
+    void *external_data,
+    size_t byte_length,
+    napi_finalize finalize_cb,
+    void *finalize_hint,
+    napi_value *result) {
+
+    napi_status external_buffer_status =
+        napi_create_external_arraybuffer(env, external_data, byte_length, finalize_cb, finalize_hint, result);
+
+    if (external_buffer_status == NAPI_NO_EXTERNAL_BUFFER_ENUM_VALUE) {
+
+        // The external buffer is disabled, manually copy the external_data into Node
+        void *napi_buf_data = NULL;
+        napi_status create_arraybuffer_status = napi_create_arraybuffer(env, byte_length, &napi_buf_data, result);
+
+        if (create_arraybuffer_status != napi_ok) {
+            AWS_NAPI_LOGF_ERROR(
+                "napi_create_arraybuffer (in aws_napi_create_external_arraybuffer) failed with : %s",
+                aws_napi_status_to_str(create_arraybuffer_status));
+            return create_arraybuffer_status;
+        }
+
+        memcpy(napi_buf_data, external_data, byte_length);
+
+        // As the data has been copied into the Node, invoke the finalize callback to make sure the
+        // data is released.
+        finalize_cb(env, finalize_hint, finalize_hint);
```

Comment:
Debatable: Currently I do not invoke `finalize_cb` if the creation failed. The function could also call the `finalize_cb` when the arraybuffer creation failed to avoid memory leak. Not sure which way is a better though.

Reply:
this is correct. You shouldn't invoke the finalizer if you're reporting an error. It's up to whatever calls aws_napi_create_external_arraybuffer() to ensure memory's cleaned up in the case of failure.